### PR TITLE
Allow tilde (~) in url parameters

### PR DESCRIPTION
**Pull request by huzaifa-99**
### Fixed Issues
$ [Expensify/App/issues/18564](https://github.com/Expensify/App/issues/18564)

# Tests
Verify if `~` works in query params

1. Send a valid url
2. Verify that the sent url appears as a valid url(blue and underlined)
3. Verify hovering over the url shows the tooltip in Web and Desktop
4. Verify that Right click (Web and Desktop) or long press (mWeb and native) shows Copy Url to Clipboard context menu
5. Edit the sent url
6. Verify steps 2-4

**Examples:**
```
https://example.com/search?q=foobar~
https://example.com/search?q=~foobar
https://example.com/?~parameter~
https://example.com/?parameter~
https://example.com/?~parameter
```

# QA
Same as tests

### Screenshots/Videos
<details>
<summary>Web</summary>

#### Safari

https://github.com/Expensify/expensify-common/assets/68777211/6a04f557-43b8-4672-8882-7058119f4ec5

#### Chrome

https://github.com/Expensify/expensify-common/assets/68777211/929f8aa4-f408-4147-8edb-73de3d2aad4e

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/expensify-common/assets/68777211/c349e3b7-f0c4-4a2d-9522-1fcfb19412b8

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/expensify-common/assets/68777211/e1deff7d-4dd8-4018-84ee-c642956cede6

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/expensify-common/assets/68777211/476985cf-a328-4a01-9139-3224745b57d3

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/expensify-common/assets/68777211/e407e70b-dae8-4190-9ece-d19691375a64

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/expensify-common/assets/68777211/d3fca658-39fa-491d-9946-e0bc1126eeab

</details>

### CODE REVIEW

**Review by sobitneupane — COMMENTED** [review]
@huzaifa-99 Can you please add test video as well? 

Along with the list of valid urls, please add following steps in Tests:

1. Send a valid url
2. Verify that the sent url appears as a valid url(blue and underlined)
3. Verify hovering over the url shows the tooltip in Web and Desktop
4. Verify that Right click (Web and Desktop) or long press  (mWeb and native) shows Copy Url to Clipboard context menu
5. Edit the sent url
6. Verify steps 2-4

File: lib/Url.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import TLD_REGEX from './tlds';
 
 const ALLOWED_PORTS = '([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])';
-const URL_PROTOCOL_REGEX = '((ht|f)tps?:\\/\\/)';
+const URL_PROTOCOL_REGEX = '((ht|f)tps?:[\\/|\\\\]{2})';
```
Comment (sobitneupane) [inline lib/Url.js:4]: Are you sure it is the right way to do it?

I am not an expert on regex. But I think you might have to use `(` not `[` with disjunction. I think something like
`(?:\\/{2}|\\\\{2})` should do the job.

**Comment by sobitneupane** [comment]
Nope. I don't mean that. I mean some other changes might be required within this repo (probably [here](https://github.com/Expensify/expensify-common/blob/3a72d6ca2421f4c879b73edb84ea7ebc622e8309/lib/ExpensiMark.js#L109)) or in Expensify/App repo.

Did you find any issue in the PR? At the present state, I don't think PR will pass [these tests](https://github.com/Expensify/expensify-common/pull/533#pullrequestreview-1422336434).

**Comment by sobitneupane** [comment]
>```
> # Trailing backslash after the TLD part
> https://www.example.com/
> ```

>With backslashes after the protocol or the trailing backslash after the TLD, the api responds like this. Notice that the anchor tag in the html key has no href prop.
Oops. Looks like 

> I also tried turning off internet. The failing urls (with backslash in domain part and trailing backslash after TLD) work when not connected to the internet, when connection is restored the api updates and it breaks the links for urls that have backslash after protocol part


Oh yes. Looks like backend is returning anchor without href. 

cc: @PauloGasparSv

File: lib/Url.js (hunk 1 above)
Comment (sobitneupane) [reply lib/Url.js:4]: Your change will accept `https:/\google.com` and `https:||google.com` as valid url not only `https:\\google.com`

**Comment by PauloGasparSv** [comment]
I'm still trying to find time today to take a look at the API and into [this](https://github.com/Expensify/expensify-common/pull/533#issuecomment-1545703681).

> If we want to follow slack then we can remove the backslash changes

That's really good to know! Whatsapp does allow `\` in the URLS though.

Taking in consideration that:

1. The original issue's intent which isn't specifically to make `\` and `\\` work but instead only `~` (this was an optional change we wanted to add)
2. Slack and GH don't recognize that `/` as part of the URL so IMO it's ok to match that behavior (despite Whatsapp allowing it).
3. The API error. I don't have a lot of time to look into this : /

I think **we shouldn't bother with making `\` work** and only fix `~` here. 

Does that sound good @sobitneupane @huzaifa-99?
That means we'll not add tests for `/` and we'll remove it from the regex changes here right?

**Comment by sobitneupane** [comment]
@huzaifa-99 Please let me know it is ready for review. Looks like you need to remove `\\` related stuffs and update your Test steps to include test steps mentioned [here](https://github.com/Expensify/expensify-common/pull/533#pullrequestreview-1422336434).

**Review by sobitneupane — APPROVED** [review]
Changes look good and test well.

### Screenshots/Videos
<details>
<summary>Web</summary>
Screenshot 2023-05-15 at 15 25 08

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Screenshot 2023-05-15 at 15 34 11

</details>

<details>
<summary>Mobile Web - Safari</summary>

Screenshot 2023-05-15 at 15 35 20


</details>

<details>
<summary>Desktop</summary>

Screenshot 2023-05-15 at 15 27 15

</details>

<details>
<summary>iOS</summary>


</details>

<details>
<summary>Android</summary>
Screenshot 2023-05-15 at 16 37 29


</details>

**Comment by PauloGasparSv** [comment]
Oh the first commit here isn't signed so I can't merge this! Checking what to do here

image
image

**Comment by PauloGasparSv** [comment]
Don't worry @huzaifa-99! 

> Update: I just found out there is an option to sign previous commits but never done it. Will try and update here soon

Thks for that!!!! I also searched a bit and asked for help internally, do you mind trying to:
- Run `git -c core.editor=true rebase --exec 'git commit --amend --no-edit -n -S' -i $COMMIT_HASH` replacing `$COMMIT_HASH` with your commit hash. (Found many similar solutions online but the team recommended this one)
- Force push (Thks for worrying about this but it's fine do to that this time)

Can you please do that and ping us @huzaifa-99?

**Comment by sobitneupane** [comment]
I don't think you should create another PR if there is already one which includes commit latest than yours.

cc: @PauloGasparSv

**Pull request merged**